src/signTxMint.c

```diff
@@ -0,0 +1,393 @@
+// static common_tx_data_t* commonTxData = &(instructionState.signTxContext.commonTxData);	//potentially for policy and UI?
+static tx_hash_builder_t* txHashBuilder = &(instructionState.signTxContext.txPartCtx.body_ctx.txHashBuilder);
+
+static inline void CHECK_STATE(sign_tx_mint_state_t expected)
+{
+	TRACE("Mint submachine state: current %d, expected %d", subctx->state, expected);
+	VALIDATE(subctx->state == expected, ERR_INVALID_STATE);
+}
+
+static inline void advanceState()
+{
+	TRACE("Advancing mint state from: %d", subctx->state);
+
+	switch (subctx->state) {
+
+	case STATE_MINT_TOP_LEVEL_DATA:
+	TRACE("  %d %d %d %d", (int)subctx->currentAssetGroup, (int)subctx->numAssetGroups, (int)subctx->numTokens, (int)subctx->currentToken);
+		ASSERT(subctx->numAssetGroups > 0);
+		ASSERT(subctx->currentAssetGroup == 0);
+		subctx->state = STATE_MINT_ASSET_GROUP;
+		break;
+
+	case STATE_MINT_ASSET_GROUP:
+	TRACE("  %d %d %d %d", (int)subctx->currentAssetGroup, (int)subctx->numAssetGroups, (int)subctx->numTokens, (int)subctx->currentToken);
+		ASSERT(subctx->currentAssetGroup < subctx->numAssetGroups);
+
+		// we are going to receive token amounts for this group
+		ASSERT(subctx->numTokens > 0);
+		ASSERT(subctx->currentToken == 0);
+
+		subctx->state = STATE_MINT_TOKEN;
+		break;
+
+	case STATE_MINT_TOKEN:
+		// we are done with the current token group
+		ASSERT(subctx->currentToken == subctx->numTokens);
+		subctx->currentToken = 0;
+		ASSERT(subctx->currentAssetGroup < subctx->numAssetGroups);
+		subctx->currentAssetGroup++;
+
+		if (subctx->currentAssetGroup == subctx->numAssetGroups) {
+			// the whole token bundle has been received
+			subctx->state = STATE_MINT_CONFIRM;
+		} else {
+			subctx->state = STATE_MINT_ASSET_GROUP;
+		}
+		break;
+
+	case STATE_MINT_CONFIRM:
+		subctx->state = STATE_MINT_FINISHED;
+		break;
+
+	default:
+		ASSERT(false);
+	}
+
+	TRACE("Advancing mint state to: %d", subctx->state);
+}
+
+static void signTxMint_handleTopLevelDataAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		// safety checks
+		CHECK_STATE(STATE_MINT_TOP_LEVEL_DATA);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	TRACE_BUFFER(wireDataBuffer, wireDataSize);
+	{
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+		VALIDATE(view_remainingSize(&view) >= 4, ERR_INVALID_DATA);
+		uint32_t numAssetGroups = parse_u4be(&view);
+		TRACE("num asset groups %u", numAssetGroups);
+		VALIDATE(numAssetGroups <= OUTPUT_ASSET_GROUPS_MAX, ERR_INVALID_DATA);
+
+		STATIC_ASSERT(OUTPUT_ASSET_GROUPS_MAX <= UINT16_MAX, "wrong max token groups");
+		ASSERT_TYPE(subctx->numAssetGroups, uint16_t);
+		subctx->numAssetGroups = (uint16_t) numAssetGroups;
+
+		VALIDATE(view_remainingSize(&view) == 0, ERR_INVALID_DATA);
+	}
+	txHashBuilder_addMint_topLevelData(txHashBuilder, subctx->numAssetGroups);
+	subctx->mintSecurityPolicy = policyForSignTxMintDefault();	//TODO not sure how to proceed here
```

Comment:
I don't understand that. If not kept in a global state, then where do we need a value at all? Right now it's set in one place and used in two others, but that was just a copy of Outputs.

Reply:
Usually, the policy is used to decide the starting step for a ui state machine; a local variable to store the policy is enough for that. For outputs, exceptionally, we apply the policy for individual tokens, too (therefore we need a variable in global context to store it between APDUs). For mint, we do need that, because we are always going to show it, but since the global context uses unions, it probably does not affect total context size at all (please check). So I would prefer to keep the code as close to outputs as possible, i.e. having the global variable even if it was actually needed only here.